akka-docs-dev/rst/scala/stream-customize.rst

```diff
@@ -129,10 +135,144 @@ Finally, there are two methods available for convenience to complete the stage a
 * ``failStage(exception)`` is equivalent to failing all output ports and cancelling all input ports.
 
 
+In some cases it is inconvenient and error prone to react on the regular state machine events with the
+signal based API described above. For those cases there is a API which allows for a more imperative sequencing
+of actions which will greatly simplify some use cases at the cost of some extra allocations. The difference
+between the two APIs could be described as that the first one is signal driven from the outside, while this API
+is more active and drives it surroundings.
+
+The operations of this part of the :class:``GraphStage`` API are:
+
+* ``emit(out, elem)`` and ``emitMultiple(out, Iterable(elem1, elem2))`` replaces the ``OutHandler`` with s handler that emit
+  one or more elements when there is demand, and then reinstall the current handlers
+* ``read(in)(andThen)`` and ``readN(in, n)(andThen)`` replaces the handlers that will
+* ``abortEmitting`` and ``abortReading`` which will cancel an ongoing emit or read
+
+An example of how this API simplifies a stage can be found below in the second version of the :class:``Duplicator``.
+
+
+Custom linear processing stages using GraphStage
+------------------------------------------------
+
+Graph stages allows for custom linear processing stages through letting them
+have one input and one output and using :class:`FlowShape` as their shape.
+
+Such a stage can be illustrated as a box with two flows as it is
+seen in the illustration below. Demand flowing upstream leading to elements
+flowing downstream.
+
+|
+
+.. image:: ../images/graph_stage_conceptual.png
+   :align: center
+   :width: 500
+
+|
+
+
+To illustrate these concepts we create a small :class:`GraphStage` that implements the ``map`` transformation.
+
+|
+
+.. image:: ../images/graph_stage_map.png
+   :align: center
+   :width: 300
+
+|
+
+Map calls push(out) from the onPush() handler and it also calls pull() form the onPull handler resulting in the
+conceptual wiring above, and fully expressed in code below:
+
+.. includecode:: code/docs/stream/GraphStageDocSpec.scala#one-to-one
+
+Map is a typical example of a one-to-one transformation of a stream where
+demand is passed along upstream elements passed on downstream.
+
+To demonstrate a many-to-one stage we will implement
+filter. The conceptual wiring of ``Filter`` looks like this:
+
+|
+
+.. image:: ../images/graph_stage_filter.png
+   :align: center
+   :width: 300
+
+|
+
+
+As we see above, if the given predicate matches the current element we are propagating it downwards, otherwise
+we return the "ball" to our upstream so that we get the new element. This is achieved by modifying the map
```

Comment:
since we switched off smart quotes, please use “…” (`option-[` and `shift-option-[` on an English Mac keyboard)